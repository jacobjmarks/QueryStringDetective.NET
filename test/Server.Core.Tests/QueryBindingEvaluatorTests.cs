namespace Server.Core.Tests;

public class QueryBindingEvaluatorTests
{
    public static readonly TheoryData<string> TestCases = new()
    {
        "",
        "q=",
        "q=c",
        "q=foo",
        "q=a&q=b&q=c",
        "q=null",
        "q=0",
        "q=1",
        "q=-1",
        "q=true",
        "q=false",
        "q=yes",
        "q=no",
        "q=1&q=2&q=3",
        "q=1&q=2&q=.3",
        "q=-1&q=0&q=1",
        $"q={sbyte.MinValue}",
        $"q={sbyte.MaxValue}",
        $"q={byte.MaxValue}",
        $"q={short.MinValue}",
        $"q={short.MaxValue}",
        $"q={ushort.MaxValue}",
        $"q={int.MinValue}",
        $"q={int.MaxValue}",
        $"q={long.MinValue}",
        $"q={long.MaxValue}",
        $"q={ulong.MaxValue}",
        $"q={Uri.EscapeDataString($"{double.MinValue}")}",
        $"q={Uri.EscapeDataString($"{double.MaxValue}")}",
        $"q={decimal.MinValue}",
        $"q={decimal.MaxValue}",
        $"q={Uri.EscapeDataString($"{float.MinValue}")}",
        $"q={Uri.EscapeDataString($"{float.MaxValue}")}",
    };

    [Theory]
    [MemberData(nameof(TestCases))]
    public async Task AssertBinding(string queryString)
    {
        using var bindingEvaluator = new QueryBindingEvaluator();
        var results = await bindingEvaluator.EvaluateAsync("?" + queryString);

        var content = $"Input: {queryString}"
            + "\n\nBindings:\n";

        foreach (var result in results)
        {
            content += "\n" + result.Type;
            content += "\n  " + string.Join("\n  ", result.Results
                .Select(r => $"{r.Key}\n    {(
                    r.Value.IsErroneous
                    ? $"e = {r.Value.Error.Message} ({r.Value.Error.Detail})"
                    : $"r = {r.Value.Result}")}"));

            content += "\n";
        }

        content.Should().MatchSnapshot(SnapshotNameExtension.Create(ToValidFileName(queryString)));
    }

    private static string ToValidFileName(string value)
    {
        foreach (var invalidChar in Path.GetInvalidFileNameChars())
            value = value.Replace(invalidChar, '_');

        return value;
    }
}