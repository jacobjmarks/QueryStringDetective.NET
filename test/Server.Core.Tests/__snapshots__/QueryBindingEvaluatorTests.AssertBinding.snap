Input: 

Bindings:

string
  MinimalApis
    r = null
  Controllers
    r = null

string?
  MinimalApis
    r = null
  Controllers
    r = null

string[]
  MinimalApis
    r = []
  Controllers
    r = []

char
  MinimalApis
    e = 400 Bad Request (Required parameter "char q" was not provided from query string.)
  Controllers
    r = "\u0000"

char?
  MinimalApis
    r = null
  Controllers
    r = null

char[]
  MinimalApis
    r = []
  Controllers
    r = []

int
  MinimalApis
    e = 400 Bad Request (Required parameter "int q" was not provided from query string.)
  Controllers
    r = 0

int?
  MinimalApis
    r = null
  Controllers
    r = null

int[]
  MinimalApis
    r = []
  Controllers
    r = []

double
  MinimalApis
    e = 400 Bad Request (Required parameter "double q" was not provided from query string.)
  Controllers
    r = 0

double?
  MinimalApis
    r = null
  Controllers
    r = null

double[]
  MinimalApis
    r = []
  Controllers
    r = []

float
  MinimalApis
    e = 400 Bad Request (Required parameter "float q" was not provided from query string.)
  Controllers
    r = 0

float?
  MinimalApis
    r = null
  Controllers
    r = null

float[]
  MinimalApis
    r = []
  Controllers
    r = []

decimal
  MinimalApis
    e = 400 Bad Request (Required parameter "decimal q" was not provided from query string.)
  Controllers
    r = 0

decimal?
  MinimalApis
    r = null
  Controllers
    r = null

decimal[]
  MinimalApis
    r = []
  Controllers
    r = []

bool
  MinimalApis
    e = 400 Bad Request (Required parameter "bool q" was not provided from query string.)
  Controllers
    r = false

bool?
  MinimalApis
    r = null
  Controllers
    r = null

bool[]
  MinimalApis
    r = []
  Controllers
    r = []

uint
  MinimalApis
    e = 400 Bad Request (Required parameter "uint q" was not provided from query string.)
  Controllers
    r = 0

uint?
  MinimalApis
    r = null
  Controllers
    r = null

uint[]
  MinimalApis
    r = []
  Controllers
    r = []

long
  MinimalApis
    e = 400 Bad Request (Required parameter "long q" was not provided from query string.)
  Controllers
    r = 0

long?
  MinimalApis
    r = null
  Controllers
    r = null

long[]
  MinimalApis
    r = []
  Controllers
    r = []

ulong
  MinimalApis
    e = 400 Bad Request (Required parameter "ulong q" was not provided from query string.)
  Controllers
    r = 0

ulong?
  MinimalApis
    r = null
  Controllers
    r = null

ulong[]
  MinimalApis
    r = []
  Controllers
    r = []

sbyte
  MinimalApis
    e = 400 Bad Request (Required parameter "sbyte q" was not provided from query string.)
  Controllers
    r = 0

sbyte?
  MinimalApis
    r = null
  Controllers
    r = null

sbyte[]
  MinimalApis
    r = []
  Controllers
    r = []

short
  MinimalApis
    e = 400 Bad Request (Required parameter "short q" was not provided from query string.)
  Controllers
    r = 0

short?
  MinimalApis
    r = null
  Controllers
    r = null

short[]
  MinimalApis
    r = []
  Controllers
    r = []

ushort
  MinimalApis
    e = 400 Bad Request (Required parameter "ushort q" was not provided from query string.)
  Controllers
    r = 0

ushort?
  MinimalApis
    r = null
  Controllers
    r = null

ushort[]
  MinimalApis
    r = []
  Controllers
    r = []

byte
  MinimalApis
    e = 400 Bad Request (Required parameter "byte q" was not provided from query string.)
  Controllers
    r = 0

byte?
  MinimalApis
    r = null
  Controllers
    r = null

byte[]
  MinimalApis
    r = ""
  Controllers
    r = null
